### PR TITLE
make distributed downstream pageable

### PR DESCRIPTION
**Pull request by seut**

### CODE REVIEW

File: sql/src/main/java/io/crate/executor/transport/distributed/DistributingDownstream.java [hunk 1]
```diff
@@ -76,36 +88,61 @@ public synchronized boolean setNextRow(Row row) {
...
+            DistributedResultRequest request = downstream.request;
+            Deque<Bucket> queue = downstream.bucketQueue;
+            int size = queue.size();
+            if (size > 0) {
+                request.rows(queue.poll());
+            } else {
+                request.rows(Bucket.EMPTY);
+            }
+            if (size > 1 || remainingUpstreams.get() > 0) {
+                request.isLast(false);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/distributed/DistributingDownstream.java]: why not  just call request.isLast() with the condition directly?

Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/distributed/DistributingDownstream.java]: good point

File: sql/src/main/java/io/crate/jobs/PageDownstreamContext.java [hunk 2]
```diff
@@ -95,10 +96,43 @@ public void setBucket(int bucketIdx, Bucket rows, boolean isLast, PageConsumeLis
...
+                            for (PageConsumeListener listener : listeners) {
+                                listener.finish();
+                            }
+                            listeners.clear();
+                        }
+                    }
+                });
+            }
+
+            // set all exhausted upstream futures
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/jobs/PageDownstreamContext.java]: why is this necessary?

File: sql/src/test/java/io/crate/integrationtests/QueryThenFetchIntegrationTest.java [hunk 3]
```diff
@@ -87,11 +87,41 @@ public void testPushBasedQTF() throws Exception {
...
-        execute("select * from t limit ?", new Object[] {Constants.PAGE_SIZE + 10000 });
+        execute("select * from t limit ?", new Object[]{Constants.PAGE_SIZE + 10000});
         assertThat(response.rowCount(), is(1L));
     }
 
     @Test
+    public void testPushBasedQTFWithPaging() throws Exception {
+        // insert more docs than PAGE_SIZE and query at least 2 times of it to trigger push of
+        // at least 2 pages
+        int docCount = (Constants.PAGE_SIZE * 2) + 2;
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/QueryThenFetchIntegrationTest.java]: I think the PAGE_SIZE was increased to 1mio,  does this test still run in an acceptable time?

File: sql/src/main/java/io/crate/jobs/PageDownstreamContext.java (hunk 2 above)
Comment (seut) [inline sql/src/main/java/io/crate/jobs/PageDownstreamContext.java]: because upstreams are not sending (useless) requests if they are exhausted, they are finished. but futures still exists (page buckets) and downstream is waiting for them

**Comment by mfussenegger** [comment]
otherwise ack

**Comment by mfussenegger** [comment]
ack
